File: OOP3Behav7.cs
// File: "OOP3Behav7"
using PT4;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PT4Tasks
{
    public class MyTask : PT
    {
        public abstract class Aggregate
        {
            public abstract Iterator CreateIterator();
        }

        // Implement the ConcreteAggregateA, ConcreteAggregateB
        //   and ConcreteAggregateC descendant classes
        public class ConcreteAggregateA : Aggregate
        {
            public ConcreteAggregateA(int data) => this.data = data;
            public override Iterator CreateIterator() => new ConcreteIteratorA(this);
            public int data { get; set; }
            public int divied { get; set; } = 1;
        }

        public class ConcreteAggregateB : Aggregate
        {
            public ConcreteAggregateB(string data) => this.data = data;
            public override Iterator CreateIterator() => new ConcreteIteratorB(this);
            public string data { get; set; }
            public int index { get; set; } = 0;
        }

        public class ConcreteAggregateC : Aggregate
        {
            public ConcreteAggregateC(int[] data) => this.data = data;
            public override Iterator CreateIterator() => new ConcreteIteratorC(this);
            public int[] data { get; set; }
            public int index { get; set; } = 0;
            public int divied { get; set; } = 1;
        }
        public abstract class Iterator
        {
            public abstract void First();
            public abstract void Next();
            public abstract bool IsDone();
            public abstract int CurrentItem();
        }

        // Implement the ConcreteIteratorA, ConcreteIteratorB
        //   and ConcreteIteratorC descendant classes
        public class ConcreteIteratorA : Iterator
        {
            public ConcreteIteratorA(ConcreteAggregateA aggr) => this.a = aggr;
            public override void First() => a.divied = 1;
            public override void Next() => a.divied *= 10;
            public override bool IsDone()
                => a.data != 0 && a.data / a.divied == 0 || a.data == 0 && a.divied == 10;
            public override int CurrentItem() => Math.Abs(a.data / a.divied % 10);

            private ConcreteAggregateA a;
        }

        public class ConcreteIteratorB : Iterator
        {
            public ConcreteIteratorB(ConcreteAggregateB a) => this.a = a;

            public override void First()
            {
                a.index = a.data.Length - 1;
                while (a.index >= 0 && !char.IsDigit(a.data[a.index]))
                {
                    a.index--;
                }
            }

            public override void Next()
            {
                a.index--;
                while (a.index >= 0 && !char.IsDigit(a.data[a.index]))
                {
                    a.index--;
                }
            }

            public override bool IsDone() => a.index < 0;
            public override int CurrentItem() => a.data[a.index] - '0';

            private ConcreteAggregateB a;
        }

        public class ConcreteIteratorC : Iterator
        {
            public ConcreteIteratorC(ConcreteAggregateC a) => this.a = a;

            public override void First()
            {
                a.index = a.data.Length - 1;
                a.divied = 1;
            }

            public override void Next()
            {
                a.divied *= 10;
                int data = a.data[a.index];
                int divied = a.divied;
                if (data != 0 && data / divied == 0 || data == 0 && divied == 10)
                {
                    a.index--;
                    a.divied = 1;
                }
            }

            public override bool IsDone() => a.index < 0;
            public override int CurrentItem() => Math.Abs(a.data[a.index] / a.divied % 10);

            private ConcreteAggregateC a;
        }
        public static void Solve()
        {
            Task("OOP3Behav7");
            int num = GetInt();
            var a = new Aggregate[num];
            for (int i = 0; i < num; i++)
            {
                char type = GetChar();
                if (type == 'A')
                {
                    int data = GetInt();
                    a[i] = new ConcreteAggregateA(data);
                }
                else if (type == 'B')
                {
                    string data = GetString();
                    a[i] = new ConcreteAggregateB(data);
                }
                else
                {
                    int n = GetInt();
                    int[] data = new int[n];
                    for (int j = 0; j < n; j++)
                    {
                        int number = GetInt();
                        data[j] = number;
                    }
                    a[i] = new ConcreteAggregateC(data);
                }
            }

            foreach (var i in a.Reverse())
            {
                var digit = new List<int>();
                var it = i.CreateIterator();
                for (it.First(); !it.IsDone(); it.Next())
                {
                    digit.Add(it.CurrentItem());
                }
                Put(digit.Sum());
                foreach (var d in digit)
                {
                    Put(d);
                }

            }
        }
    }
}

File: OOP3Behav7.cpp
#include "pt4.h"
#include <numeric>
using namespace std;

class Iterator
{
public:
    virtual void First() = 0;
    virtual void Next() = 0;
    virtual bool IsDone() = 0;
    virtual int CurrentItem() = 0;
};

class Aggregate
{
public:
    virtual Iterator* CreateIterator() = 0;
};

// Implement the ConcreteAggregateA, ConcreteAggregateB
//   and ConcreteAggregateC descendant classes

class ConcreteAggregateA : public Aggregate
{
public:
    ConcreteAggregateA(int data) : data(data) {}
    Iterator *CreateIterator() override;
    int data;
    int divied = 1;
};

class ConcreteAggregateB : public Aggregate
{
public:
    ConcreteAggregateB(string data) : data(data) {}
    Iterator *CreateIterator() override;
    string data;
    int index = 0;
};

class ConcreteAggregateC : public Aggregate
{
public:
    ConcreteAggregateC(vector<int> data) : data(data) {}
    Iterator *CreateIterator() override;
    vector<int> data;
    int index = 0;
    int divied = 1;
};
// Implement the ConcreteIteratorA, ConcreteIteratorB
//   and ConcreteIteratorC descendant classes
class ConcreteIteratorA : public Iterator
{
public:
    ConcreteIteratorA(ConcreteAggregateA *aggr) : aggr(aggr) {}
    virtual void First() override { aggr->divied = 1; }
    virtual void Next() override { aggr->divied *= 10; }
    virtual bool IsDone() override { return aggr->data && aggr->data / aggr->divied == 0 || !aggr->data && aggr->divied == 10; }
    virtual int CurrentItem() override { return abs(aggr->data / aggr->divied % 10); }

private:
    ConcreteAggregateA *aggr;
};

class ConcreteIteratorB : public Iterator
{
public:
    ConcreteIteratorB(ConcreteAggregateB *aggr) : aggr(aggr) {}
    virtual void First() override
    {
        aggr->index = aggr->data.length() - 1;
        while (aggr->index >= 0 && !isdigit(aggr->data[aggr->index]))
            aggr->index--;
    }
    virtual void Next() override
    {
        aggr->index--;
        while (aggr->index >= 0 && !isdigit(aggr->data[aggr->index]))
            aggr->index--;
    }
    virtual bool IsDone() override { return aggr->index < 0; }
    virtual int CurrentItem() override { return aggr->data[aggr->index] - '0'; }

private:
    ConcreteAggregateB *aggr;
};

class ConcreteIteratorC : public Iterator
{
public:
    ConcreteIteratorC(ConcreteAggregateC *aggr) : aggr(aggr) {}
    virtual void First() override
    {
        aggr->index = aggr->data.size() - 1;
        aggr->divied = 1;
    }
    virtual void Next() override;
    virtual bool IsDone() override { return aggr->index < 0; }
    virtual int CurrentItem() override { return abs(aggr->data[aggr->index] / aggr->divied % 10); }

private:
    ConcreteAggregateC *aggr;
};

void ConcreteIteratorC::Next()
{
    aggr->divied *= 10;
    int data = aggr->data[aggr->index];
    int divied = aggr->divied;
    if (data && data / divied == 0 || !data && divied == 10)
    {
        aggr->index--;
        aggr->divied = 1;
    }
}

Iterator *ConcreteAggregateA::CreateIterator()
{
    return new ConcreteIteratorA(this);
}
Iterator *ConcreteAggregateB::CreateIterator()
{
    return new ConcreteIteratorB(this);
}
Iterator *ConcreteAggregateC::CreateIterator()
{
    return new ConcreteIteratorC(this);
}

void Solve()
{
    Task("OOP3Behav7");
    int n;
    pt >> n;
    vector<Aggregate *> a(n);
    for (int i = 0; i < n; i++)
    {
        char c;
        pt >> c;
        if (c == 'A')
        {
            int num;
            pt >> num;
            a[i] = new ConcreteAggregateA(num);
        }
        else if (c == 'B')
        {
            string num;
            pt >> num;
            a[i] = new ConcreteAggregateB(num);
        }
        else
        {
            int k, num;
            pt >> k;
            vector<int> v(k);
            for (int i = 0; i < k; i++)
            {
                pt >> num;
                v[i] = num;
            }
            a[i] = new ConcreteAggregateC(v);
        }
    }

    for (auto it = a.rbegin(); it != a.rend(); it++)
    {
        vector<int> v2;
        auto it2 = (*it)->CreateIterator();
        for (it2->First(); !it2->IsDone(); it2->Next())
            v2.push_back(it2->CurrentItem());
        pt << accumulate(v2.begin(), v2.end(), 0);
        for (auto i : v2)
            pt << i;
    }
}
